Ingress/controllers/nginx-third-party/controller.go

```diff
@@ -0,0 +1,237 @@
+				key, ok := values[i].(string)
+				if !ok {
+					return nil, errors.New("dict keys must be strings")
+				}
+				dict[key] = values[i+1]
+			}
+			return dict, nil
+		},
+	}
+
+	tmpl, _ := template.New("nginx.tmpl").Funcs(funcMap).ParseFiles("./nginx.tmpl")
+	rateLimiter := util.NewTokenBucketRateLimiter(0.1, 1)
+
+	known := &extensions.IngressList{}
+	defCfg := defNginxCfgParams()
+	runningCfg := make(map[string]interface{})
+
+	sslCertificates := createSSLCerts(sslDirectory)
+
+	servicesL4 := getTcpServices(kubeClient, *tcpServices)
+	if len(servicesL4) == 0 {
+		glog.Info("no tcp service defined. skipping")
+	} else {
+		glog.Infof("tcp service/s defined: %v", servicesL4)
+	}
+
+	defBackendSvc := getService(kubeClient, *defaultSvc)
+	defErrorSvc := getService(kubeClient, *customErrorSvc)
+	defDhFile := dhParamFile(sslDirectory)
+
+	// we need to configure the nginx resolver parameter
+	// http://nginx.org/en/docs/http/ngx_http_core_module.html#resolver
+	// if kube-dns is not working it will return a fatal error.
+	defNgxResolver := skydnsIPPort(kubeClient)
+
+	// we need to start nginx and terminate the process if it dies
+	go startNginx()
+
+	conf := make(map[string]interface{})
+	conf["sslCertificates"] = sslCertificates
+	conf["tcpServices"] = servicesL4
+	conf["defErrorSvc"] = defErrorSvc
+	conf["defBackend"] = defBackendSvc
+	conf["sslDHParam"] = defDhFile
+
+	go registerHandlers()
+
+	// Controller loop
+	for {
+		rateLimiter.Accept()
+		ingresses, err := ingClient.List(labels.Everything(), fields.Everything())
+		if err != nil {
+			glog.Errorf("error retrieving ingresses: %v", err)
+			continue
+		}
+
+		newNginxCfg, err := cfgFrom3rdPartyResource(kubeClient, *configurationName)
+		if err != nil {
+			glog.Errorf("error retrieving configuration: %v", err)
+			newNginxCfg = defCfg
+		}
+
+		fromMap := structs.Map(newNginxCfg)
+		toMap := structs.Map(defCfg)
+		curNginxCfg := merge(toMap, fromMap)
+
+		if val, ok := curNginxCfg["Resolver"]; ok {
+			if val == "" {
+				if defNgxResolver == "" {
+					glog.Fatalf("No custom resolver parameter configured or SkyDNS service found in the cluster.")
+				}
+
+				curNginxCfg["Resolver"] = defNgxResolver
+			}
+		} else {
+			curNginxCfg["Resolver"] = defNgxResolver
+		}
+
+		glog.V(2).Info("checking for new nginx configuration")
+		if !reflect.DeepEqual(runningCfg, curNginxCfg) {
+			glog.Info("new configuration detected. forcing nginx reload")
+		} else if reflect.DeepEqual(ingresses.Items, known.Items) {
+			glog.V(2).Info("skipping reload. No new Ingress rules or new nginx configuration")
+			continue
+		}
+
+		known = ingresses
+		runningCfg = curNginxCfg
+
+		conf["cfg"] = curNginxCfg
+		conf["ingresses"] = ingresses
+
+		if glog.V(2) {
+			b, err := json.Marshal(conf)
+			if err == nil {
+				glog.Infof("nginx configuration: %v", string(b))
+			}
+		}
+
+		if w, err := os.Create("/etc/nginx/nginx.conf"); err != nil {
+			glog.Fatalf("failed to open nginx configuration: %v", err)
```

Comment:
Just loop around logging the error? that way someone can get logs, and modify perms if they want. It's a little more graceful than a crashloop.